app/controllers/stats_controller.rb

```diff
@@ -141,7 +141,7 @@ def last_four_months_hash(association, date_attribute)
 
     association
       .where(date_attribute => min_date..max_date)
-      .group("DATE_TRUNC('month', #{date_attribute})")
+      .group("DATE_TRUNC('month', #{date_attribute}::TIMESTAMPTZ AT TIME ZONE '#{Time.zone.formatted_offset}'::INTERVAL)")
```

Comment:
Alors sur le principe je suis assez partant ; merci pour le code.

Cela dit, toutes les fois où j'ai commencé à rencontrer des problèmes de timezone dans des stats, je me suis rapidement retrouvé mieux à utiliser plutôt les petites fonctions de la gem [groupdate](https://github.com/ankane/groupdate).

Essentiellement ça permet typiquement d'écrire ça :

```suggestion
      .group_by_month(date_attribute)
```

Et ça groupe automatiquement en fonction de `Time.zone`.

Est-ce que ça te semblerait raisonnable de plutôt utiliser cette gem, plutôt que d'ajuster les truncate manuellement ? (Si ça te semble chaud, je peux m'en occuper moi-même.)

Reply:
Super suggestion. je fais ca et je poste !

Reply:
Deux modifications sémantiques où j'ai besoin de votre validation: 
La fonction *last4_month_hash* retournait toujours 3 stats dans un tableau.
Or quand je vois une fonction *last_4_month_hash*, je m'attends 
  * A ce qu'elle retourne 4 valeurs et pas 3
  * A ce qu'elle retourne un hash et pas un tableau (même si chartkick s'en accommode plus tard)
 
Qu'en pensez vous ?  
Je reste sur le comportement initial ?

Reply:
Alors j'ai voulu faire le refactor pour passer en mode 3_mois... 
Mais ca descend jusqu'à la table stat dans la base qui indique 'last_4_months'. 
Et ceci dit, pour utiliser les stats assez souvent je préfère les stats en mode 4 mois c'est beaucoup plus parlant. 
Bref 3 chemins: 
   - on laisse le nommage 4 months et on retourne 3 mois (situation actuelle)
   - on laisse le nommage 4 months et on retourne 4 mois (l'option que je préfère)
   - on migre la table stats en base et le nommage à trois mois et on retourne 3 mois ....
Votre avis ?

Reply:
Si en base on indique `last_4_months`, je dirais qu'on garde ça, oui, et qu'on ajuste le code pour que ça renvoie bien 4 mois (donc l'option que tu préfères).